### ISSUE TITLE
Unable to record/transcribe

### ISSUE THREAD

**Issue by chejmadi** [introlab/odas_web#44]
Hi,
I'm running ODAS on a Raspberry Pi with Respeaker 4-mic array and running ODAS Web on a remote laptop running WSL on Windows 10. I'm somehow just not able to save any recordings once ODAS Web is running. I check the "Determine if separated audio is recorded" box. As soon as I start speaking, the microphone array picks up my voice and the terminal on my remote machine shows that "Recorder 0 started", but then shows "Recorder 0 was false active" - it then goes "Write stream 0 full", "Holding samples" and "Recorder 0 ended". The files that get saved are either of duration -1.00 seconds like in another issue here, or 1.54 seconds long, with barely any discernible sound. Also as soon as the recorder begins, the ODAS Web GUI really slows down, and I can't even kill the process on the Raspberry Pi easily. 

Plus, the transcription function is not working either, although I'm not sure if that's due to this recorder problem or a separate issue of its own. I have an API key for Google's speech API (at least, I think so) but I'd like to know where one should download it from. I'm attaching my API key, you guys could confirm if it's correct or not. 
[SpeechTranscription-f7dd9e0aeebf.zip](https://github.com/introlab/odas_web/files/4479745/SpeechTranscription-f7dd9e0aeebf.zip)

Thanks!

**Comment by GodCed** [comment]
Hi, let's start with the recording issue.

It looks like the ODAS Web process cannot write the audio file fast enough. I never used the WSL so I can't tell if the issue comes from there. ODAS Web then hangs because the buffer are getting out of hands (the slow GUI part) and stops receiving audio samples, which leads to ODAS freezing (that's why the process is hard to kill on the Pi). What is your ODAS sink config? Maybe the sample rate is juste too high and it can't write fast enough (parsing raw audio in a javascript loop is a rather suboptimal process). 16 000 is normally a safe choice.

For the transcription part, I will refer you to the [Google Cloud documentation](https://cloud.google.com/speech-to-text/docs/libraries#setting_up_authentication) for getting your key file. However if recording is not working I'm not sure text to speech will, so I would try to sort one issue at the time.

**Comment by chejmadi** [comment]
I don't know what you mean exactly by "sink config" but the config file I run for ODAS on the RPi has the sampling rate for "raw" as 16000. However the sampling frequency for "separated" and "postfiltered" are 44100.

**Comment by GodCed** [comment]
Yes those are the ones. They "sink" data out of ODAS. Try setting them at 16000 like your RAW (and adjust the ODAW Web configuration accordingly).

**Comment by chejmadi** [comment]
Okay, that makes sense. I'll try!

**Comment by chejmadi** [comment]
Also, should they be the same as RAW even when I'm recording on the raspberry Pi itself (without ODAS Web, just saving it to a file on the Pi)?

**Comment by GodCed** [comment]
Except if your application really requires a specific sample rate they should. There is no advantage to upsampling above the RAW sample rate that I know of, because you can't "create" new information that was not present in the source signal.

**Comment by chejmadi** [comment]
I see. 
So I changed the sampling frequency to 16000, but the problem persists. The audio files have recorded but are of only 1.54 seconds length, however now I can clearly hear my voice in them. But the length problem is still there. I'm going to try and upload a screenshot of my laptop terminal.
I have no idea how "Recorder 23 was defined" turns up there. Then there's a bunch of (what I think are) Javascript messages.
![image](https://user-images.githubusercontent.com/46592147/79359953-3b0b7680-7f11-11ea-8739-18c72bfcae8d.png)
![image](https://user-images.githubusercontent.com/46592147/79360130-7d34b800-7f11-11ea-9c92-8e5d7347cd45.png)

**Comment by chejmadi** [comment]
I've also noticed that it closes the connection by itself after showing that the Write Stream is full. I haven't touched anything on the Pi end. The connection got closed somehow, even though there was no error message on the Pi's terminal. 

![image](https://user-images.githubusercontent.com/46592147/79384805-8cc5f800-7f35-11ea-9a80-32bff8743cc1.png)

![image](https://user-images.githubusercontent.com/46592147/79384920-bb43d300-7f35-11ea-85d5-11fd2960b9b8.png)

Once the connection closes, it starts outputting this on the terminal on my laptop
![image](https://user-images.githubusercontent.com/46592147/79385655-f397e100-7f36-11ea-835a-643b5d79878d.png)

At this point, two 1.54s recordings have been processed and can be played back. Two are still processing, by the looks of things. 
![image](https://user-images.githubusercontent.com/46592147/79385736-13c7a000-7f37-11ea-86e9-2eaf5e2d6bed.png)

And at the moment it seems like they aren't going to be processed. There's a request timeout.
![image](https://user-images.githubusercontent.com/46592147/79385818-2f32ab00-7f37-11ea-9af0-388d7cd01749.png)

**Comment by GodCed** [comment]
The connection closing on the Pi is expected as ODAS simply stops without any message when it can't sink in real time.

There is a mix of request timeout from Google Speech and recording buffer full in the ODAS Studio output. Can you disable the Google Speech Transcription for now it will isolate things and make the output cleaner. Also if you could get a terminal output with proper line termination it would certainly improve readability.

As for recorder up to 34, that is really strange as recorders are created in a loop from 0 to 3.

**Comment by chejmadi** [comment]
Yeah the line termination drives me nuts too. Sorry. I didn't know if I can change that. Not sure if it's an electron thing or npm thing (it was a huge pain getting those two things set up on my laptop).
Anyway I disabled the Google Speech Transcription and hey presto, the recorder started working again! I guess there has to be something wrong with the API key or something, I'll go through the link you shared. Thanks!

**Comment by GodCed** [comment]
Glad to hear it’s working. From what I seen your key file seems good, the error

> Error 14: UNAVAIBLE

Leads me to think there is a network connectivity problem from the laptop to the google API.

A quick search turned out [this thread](https://stackoverflow.com/questions/51250372/error-14-unavailable-connect-failed-with-google-speech-api).

I don’t know how networking works trough the Windows Linux subsystem but Google API seems to dislike proxies so maybe, but this is a far fetched maybe, it comes from there.
